### ISSUE TITLE
cases when not unmounting causes render failures in storybook and storyshots

### ISSUE THREAD

**Issue by jrmurad** [storybooks/storybook#1634]
In my decorator, I reset global state to empty with redux.

In each story, I dispatch actions which set the required global data for the component to be rendered.

When I use my browser's refresh while on such a story or go directly to its URL, I get an error page because it:

1. runs story fine
2. for some reason, runs the story again
3. this causes the decorator's reset to run
4. the reset clears required data from the component so it breaks

I believe the correct thing to do would be to either (a) not run the story twice or (b) unmount first before running again.

Two different hacks got this working for me.

1. Global decorator with:

```
    if (typeof document !== 'undefined') {
        ReactDOM.unmountComponentAtNode(document.getElementById('root'));
    }
```

2. Removing the guard put in for https://github.com/storybooks/react-storybook/issues/116

A related issue occurs when using storyshots. The previous story component does not unmount between stories. I was able to fix this by modifying `test-bodies.js` to unmount. I can submit a PR if this is an appropriate solution.

```
export const snapshotWithOptions = options => ({ story, context }) => {
  const storyElement = story.render(context);
  const testInstance = renderer.create(storyElement, options);
  const tree = testInstance.toJSON();
  expect(tree).toMatchSnapshot();
  testInstance.unmount();
}
```

For reference, this is how my stories look:

```
import {dispatch} from 'store';
import {reset} from 'ducks';

import SomeComponent from 'containers/SomeComponent'

storiesOf('story', module)
    .addDecorator(story => {
        dispatch(reset());
        return story();
    })
    .add('pending removal', () => {
        dispatch(...); // if data set here is required by SomeComponent, we have problems
        return <SomeComponent />;
```

**Comment by jrmurad** [comment]
Note: this is still occurring with 3.3alpha and the state workaround handles it for me.

**Comment by danielduan** [comment]
Published in 3.2.17
